tools/dashboard/control_board.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+
+import argparse
+import asyncio
+import atexit
+import logging
+import os
+
+from artiq import __artiq_dir__ as artiq_dir
+from artiq import __version__ as artiq_version
+from artiq.dashboard import applets_ccb, datasets
+from artiq.gui import log, state
+from artiq.gui.models import ModelSubscriber
+from artiq.tools import get_user_config_dir
+from jax.tools.applets.dds import DDS
+from jax.tools.applets.explorer import Explorer
+from jax.tools.applets.schedule import Schedule
+from PyQt5 import QtCore, QtGui, QtWidgets
+from qasync import QEventLoop
+from sipyco import common_args
+from sipyco.asyncio_tools import atexit_register_coroutine
+from sipyco.broadcast import Receiver
+from sipyco.pc_rpc import Client
+
+
+def get_argparser():
+    parser = argparse.ArgumentParser(description="ARTIQ Control Dashboard")
+    parser.add_argument(
+        "--version",
+        action="version",
+        version="ARTIQ v{}".format(artiq_version),
+        help="print the ARTIQ version number",
+    )
+    parser.add_argument(
+        "-s",
+        "--server",
+        default="::1",
+        help="hostname or IP of the master to connect to",
+    )
+    parser.add_argument(
+        "--port-notify",
+        default=3250,
+        type=int,
+        help="TCP port to connect to for notifications",
+    )
+    parser.add_argument(
+        "--port-control",
+        default=3251,
+        type=int,
+        help="TCP port to connect to for control",
+    )
+    parser.add_argument(
+        "--port-broadcast",
+        default=1067,
+        type=int,
+        help="TCP port to connect to for broadcasts",
+    )
+    parser.add_argument(
+        "--db-file", default=None, help="database file for local GUI settings"
+    )
+    common_args.verbosity_args(parser)
+    return parser
+
+
+class MainWindow(QtWidgets.QMainWindow):
+    def __init__(self, server):
+        QtWidgets.QMainWindow.__init__(self)
+
+        # icon = QtGui.QIcon(os.path.join(artiq_dir, "gui", "logo.svg"))
+        # self.setWindowIcon(icon)
+        self.setWindowTitle("ARTIQ Control Dashboard - {}".format(server))
+
+        qfm = QtGui.QFontMetrics(self.font())
+        self.resize(140 * qfm.averageCharWidth(), 38 * qfm.lineSpacing())
+
+        self.exit_request = asyncio.Event()
+
+    def closeEvent(self, event):
+        event.ignore()
+        self.exit_request.set()
+
+    def save_state(self):
```

Comment:
should we add a button that does this? I'm a little confused about the current use for it

Reply:
This function (and restoreState) is called when the dashboard initializes. This is a part of PyQt for restoring and saving the state of the window. Again the code is copied from artiq